### ISSUE TITLE
Support for pagination

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#538]
**Reported by @rudolfmayer on 20 Oct 2007 13:40 UTC**
Several pages in karamba do pagination and display the number of pages & total results, especially search result pages.
this means

1.) issuing a second, identical query to the DB w/o limit / offset to get the total count. this seems is btw not always done accurate, as printing mak:values from nullable fields seems to influence the result from the "real" query

2.) doing some maths to find out how many total pages there are with the current limit, and on what page we are

3.) display navigation links to first/previous/next/last page

for sure 1.) should be somehow be built in to makumba, but also the others could get support

- provide mak:totalCount() or so, that would take care of doing doing an identical query to get the count. i guess this basically sums up to replacing everything between SELECT and FROM with count(*) ?

- provide mak:currentPage() that gives you the current page, and mak:numberOfPages() that gives you the total number. obviously, both depend on the mak:totalCount() being evaluated before, which should happen automatically if needed.

the problem with 3.) is mainly with search forms to pass on the parameters, but  i think this could be solved in bug 957

Migrated-From: http://trac.makumba.org/ticket/990

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 20 Oct 2007 21:52 UTC**
> this seems is btw not always done accurate, as printing mak:values
> from nullable fields seems to influence the result from the "real" query

would be curious to learn more... nullable fields should introduce separate queries so they should not affect the main one. 

-----

i think the most important problem of makumba pagination currently is that pagination is actually done only in the external mak:list, the others return all their results and only the results relevant for the external mak:list are shown. see bug 923 comment 2 for a solution

-----

for the rest, pagination can be done easily, i would look for taglibs out there that, given the total number, the page size and the current position, generate all links, plus mak:input to change the number of items per page. in fact my suggestion is to do a separate little taglib for this:

<page:list separator=" ">
<page:linkToPageNumber/>
</page:list>
<page:linkToNext/>
<page:linkToPrevious/>

<page:formToChangeInputItemsPerPage />

all links would link to the same page with same parameters, and the form would do the same, but pass the page params as hidden inputs, and have one input for "items per page"

then:

<mak:list .... />
<page:setTotal>lastCount()</page:setTotal>
<mak:list ... offset="#{page:offset}" limit="#{page:limit}" > ...

the last 3 lines can then be written as

<mak:list ... usePagination="true" >....

in other words i suggest that the mak: taglib knows about  page but not the other way around

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 20 Oct 2007 21:55 UTC**
(In reply to comment #1)
> shown. see bug 923 comment 2 for a solution

i mean bug 923 comment 3

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 26 Oct 2007 16:05 UTC**
(In reply to comment #1)
> > this seems is btw not always done accurate, as printing mak:values
> > from nullable fields seems to influence the result from the "real" query
> 
> would be curious to learn more... nullable fields should introduce separate
> queries so they should not affect the main one. 


maybe it is not due to nullable fields, but there is something different - http://private.best.eu.org/archive/archiveBrowse.jsp?archive=6s42ik4 shows an archive that has 6 documents listed (which are i guess either orphan files or sub-folders), but they are not displayed. i think there is some more mak:value / something else done in the list that actually displays the content, that does some join or so, so they do not appear.

> -----
> i think the most important problem of makumba pagination currently is that
> pagination is actually done only in the external mak:list, the others return
> all their results and only the results relevant for the external mak:list are
> shown. see bug 923 comment 2 for a solution


i don't really get what you mean with the external mak:list.. the list that does the actual counting you mean?

> -----
> for the rest, pagination can be done easily, i would look for taglibs out there
> that, given the total number, the page size and the current position, generate
> all links, plus mak:input to change the number of items per page. in fact my
> suggestion is to do a separate little taglib for this:


indeed i found something out there - http://jsptags.com/tags/navigation/pager/index.jsp looks pretty promising.

> <page:list separator=" ">
> <page:linkToPageNumber/>
> </page:list>
> <page:linkToNext/>
> <page:linkToPrevious/>


this can be done with that taglib
 
> <page:formToChangeInputItemsPerPage />


this not, we will see.
 
> all links would link to the same page with same parameters, and the form would
> do the same, but pass the page params as hidden inputs, and have one input for
> "items per page"
> 
> then:
> 
> <mak:list .... />
> <page:setTotal>lastCount()</page:setTotal>


this is rather bad, no? cause we should rather use count(*) than do a lastCount(). also, i really think that a totalCount() in mak:list that would take the exact query and select only count(*) would be the way to go.

> <mak:list ... offset="#{page:offset}" limit="#{page:limit}" > ...
> 
> the last 3 lines can then be written as
> 
> <mak:list ... usePagination="true" >....
> 
> in other words i suggest that the mak: taglib knows about  page but not the
> other way around


well, usePagination="true" imho would mainly take care about limit & offset applying, right? or also do something with the numberOfItems thingy?

greetings
rudi

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 26 Oct 2007 18:48 UTC**
(In reply to comment #3)
> (In reply to comment #1)

> maybe it is not due to nullable fields, but there is something different -
> http://private.best.eu.org/archive/archiveBrowse.jsp?archive=6s42ik4 shows an
> archive that has 6 documents listed (which are i guess either orphan files or
> sub-folders), but they are not displayed. i think there is some more mak:value
> / something else done in the list that actually displays the content, that does
> some join or so, so they do not appear.

well, some more analysis seems to be needed
> i don't really get what you mean with the external mak:list.. the list that
> does the actual counting you mean?

the root mak:list.
if there are other mak:lists inside it, they will not be limited while talking to the db because there's no way to know how to limit them. their results irrelevant to the root mak:list of course don't make it to the served page due to the mak:list engine
> > <page:setTotal>lastCount()</page:setTotal>
> 
> 
> this is rather bad, no? cause we should rather use count(*) than do a
> lastCount(). also, i really think that a totalCount() in mak:list that would
> take the exact query and select only count(*) would be the way to go.

sure, that's bug 25.

> well, usePagination="true" imho would mainly take care about limit & offset
> applying, right? or also do something with the numberOfItems thingy?

only limit and offset

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 28 Oct 2007 18:09 UTC**
> > well, usePagination="true" imho would mainly take care about limit & offset
> > applying, right? or also do something with the numberOfItems thingy?
> 
> only limit and offset


in fact, i think the only thing that mak:list would need to know about pagination & limit/offset is a kind of "default limit", that is if there is no limit provided int he request, it would use the default limit instead of unlimited

thus, an attribute defaultLimit="..." would be sufficient

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 14 Jan 2009 20:42 UTC**
rudi implemented support for pagination
